Add binaryMessenger constructor argument to platform channels

harryterkelsen opened the pull request:
By abstracting out the calls to [BinaryMessages] the same code
can be used for platform channels on the platform side as on the
framework side.

## Description

This PR adds base classes to MethodChannel and BasicMessageChannel, which abstracts out a 'BinaryMessenger' which is responsible for actually moving the bytes across the platform divide. The existing Channel implementations use BinaryMessages to implement their BinaryMessenger. The "dual" channel, which sends messages from the platform side to the framework side, would use a different BinaryMessenger.

## Related Issues

- #29629

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]). Link to flutter-announce email: https://groups.google.com/d/msg/flutter-announce/oWP9TCKys2E/1pExn2RVBQAJ
- [ ] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

harryterkelsen committed:
Add super-classes to MethodChannel and BasicMessageChannel.

By abstracting out the calls to [BinaryMessages] the same code
can be used for platform channels on the platform side as on the
framework side.

amirh commented:
Curious what do you think about the alternative of using the same mechanism `setMockMessageHandler` uses?

harryterkelsen committed:
Fix formatting issues

harryterkelsen committed:
More formatting

hterkelsen commented:
After several discussions, I think we plan to support 2 approaches for plugins:

1. You write a web plugin in Dart that is pretty a Dart port of the Android-specific code. In this approach, you need a platform channel that can send messages back to the framework side. I think it is cleaner to have Platform-side implementations of Channel than to register mock handlers on BinaryMessages.

2. In the plugin, you have an `if (const bool('dart.library.html')) { ... }` where you have web-specific logic. This avoids channels altogether.

harryterkelsen committed:
Remove trailing whitespace

harryterkelsen committed:
Merge branch 'master' of github.com:flutter/flutter into flutter_channels_2

ClassyRemade reviewed (commented):
My name was not my intention. It's out there now!

hterkelsen commented:
Ping!

goderbauer commented:
@amirh @collinjackson @bparrishMines Can somebody review this?

hterkelsen commented:
@Hixie FYI this PR incorporates the changes we talked about in our meeting a few weeks ago.

yjbanov commented on packages/flutter/lib/src/services/binary_messenger.dart:
```diff
@@ -0,0 +1,42 @@
+  /// The handler's return value, if non-null, is sent as a response, unencoded.
+  void setMessageHandler(String channel, Future<ByteData> handler(ByteData message));
+
+  /// Set a mock callback for intercepting messages from the `send*` methods on
```
There's only one `send*` method on this class, so I think the dartdoc can be more specific.

hterkelsen replied:
Done

yjbanov commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -28,24 +29,27 @@ import 'platform_messages.dart';
-  /// Neither [name] nor [codec] may be null.
-  const BasicMessageChannel(this.name, this.codec);
+  /// None of [name], [codec], or [binaryMessenger] may be null.
+  const BaseMessageChannel(this.name, this.codec, this.binaryMessenger);
```
Can we assert that the inputs are not null?

hterkelsen replied:
Done.

yjbanov commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -28,24 +29,27 @@ import 'platform_messages.dart';
   /// The message codec used by this channel, not null.
   final MessageCodec<T> codec;
 
+  /// The messenger which sends the bytes for this channel, not null.
```
nit: the ", not null" makes it sound like you are trying to say that the messenger does not send nulls.

hterkelsen replied:
Done.

yjbanov commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -512,3 +540,21 @@ class EventChannel {
   }
 }
+
+/// A [BinaryMessenger] which uses [BinaryMessages] to send platform messages.
```
I think it would be useful to say where those messages will go, i.e. the platform plugins.

hterkelsen replied:
Done.

harryterkelsen committed:
Merge branch 'master' of github.com:flutter/flutter into flutter_channels_2

harryterkelsen committed:
Respond to review comments

harryterkelsen committed:
Merge branch 'master' into flutter_channels_2

Hixie commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -27,25 +28,31 @@ import 'platform_messages.dart';
-  ///
-  /// Neither [name] nor [codec] may be null.
-  const BasicMessageChannel(this.name, this.codec);
+/// See: <https://flutter.io/platform-channels/>
```
merge conflict in the url here

Hixie replied:
(should be .dev)

Hixie commented:
This ends up being a bit convoluted, with BinaryMessages, BinaryMessenger, MethodChannel, BaseMethodChannel, BasicMessageChannel, and BaseMessageChannel. I'm not sure I'd be able to draw an accurate diagram of which class inherited from which, and which instantiated which, and which was static vs which was instantiatable, without referencing the code.

Could we instead just make BinaryMessages non-static? i.e. refactor it so that you call through to an singleton instance of that class, much as we do with things like the ImageCache? Then for the framework side the binding could set it up, and on the plugin side the plugin could set it up.

hterkelsen commented:
That might not work for the web case. On the web we compile all of the code (platform side and client side) into one executable, so we wouldn't be able to have the two different compilations set up the `binaryMessages` singleton in two different ways.

What do you think about changing this CL so that instead of introducing a super-class, we just add an optional `binaryMessenger` argument to `MethodChannel` which defaults to one which uses `BinaryMessages`. On the platform side users can either use a normal method channel with their own `binaryMessenger` or we can supply a platform-side subclass of `MethodChannel` which has a `binaryMessenger` that sends data back to the framework side.

harryterkelsen committed:
Merge branch 'master' of github.com:flutter/flutter into flutter_channels_2

harryterkelsen committed:
Instead of a superclass, just add an optional constructor argument.

dnfield commented:
@hterkelsen @Hixie what's the path forward here?  Are there more changes needed to this, or is it ready to land?

Hixie commented:
Sorry, haven't been checking my bugmail much recently.

The new PR is much closer to something reasonable, I think. Can we go further and just have BinaryMessages be an instance instead of static? Then we don't need both BinaryMessenger and BinaryMessages, we can just have the interface (public), and the default implementation (also public, but instance methods instead of static methods, and used as the default value to the arguments added in the current PR).

dnfield commented:
@hterkelsen - do you plan to continue with this?  @Hixie - are these concerns meant to block landing this?

At this point the PR has been open for over a month - if we don't think we can keep making progress forward with this approach perhaps we should close it.

Hixie commented:
> are these concerns meant to block landing this?

yes

dnfield reviewed (changes requested):
(making this easier to see it needs changes when doing triage/review - see Hixie's comments)

hterkelsen commented:
Yes, I will update this PR shortly (I am at the Dart Summit this week)

harryterkelsen committed:
Merge branch 'master' into flutter_channels_2

hterkelsen commented:
Making `BinaryMessages` an instance instead of static is a breaking change. Many plugin tests make use of `BinaryMessages.handlePlatformMessage` and `BinaryMessages.setMockMethodHandler`. Unless we propose naming the default instance of `BinaryMessenger` `BinaryMessages` those tests will need to be migrated. Is it worth it?

Hixie commented:
Can we have the static methods redirect to the default instance?

Hixie commented:
I guess there's no way to create a static method with the same name as an instance method so maybe not...

Hixie commented:
we could rename the members while we're at it, then have the old names redirect to the instance methods on the default instance, and simultaneously deprecate them so people know to migrate their code.

harryterkelsen committed:
Merge branch 'master' into flutter_channels_2

harryterkelsen committed:
Deprecate BinaryMessages and replace it with a default instance.

hterkelsen commented:
PTAL @dnfield and @Hixie 

Instead of renaming the methods, I deprecated `BinaryMessages` wholesale and replaced it with a default instance of `BinaryMessenger`. The `BinaryMessages` static methods forward to the default instance. Please let me know if this is what you had in mind.

Hixie commented on packages/flutter/lib/src/services/binary_messenger.dart:
```diff
@@ -0,0 +1,153 @@
+
+import 'package:flutter/foundation.dart';
+
+typedef _MessageHandler = Future<ByteData> Function(ByteData message);
```
this could be public, since you use the same signature in public APIs.

hterkelsen replied:
Done.

Hixie commented on packages/flutter/lib/src/services/binary_messenger.dart:
```diff
@@ -0,0 +1,153 @@
+  const _DefaultBinaryMessenger._();
+
+  // Handlers for incoming messages from platform plugins.
+  // Note: This is static so that this class can have a const constructor.
```
here are below: no need to say "Note:". See https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-empty-prose

hterkelsen replied:
Done.

Hixie commented on packages/flutter/lib/src/services/platform_messages.dart:
```diff
@@ -6,95 +6,50 @@ import 'dart:async';
 ///  * [EventChannel], which provides platform communication using event streams.
 ///  * <https://flutter.dev/platform-channels/>
+@Deprecated('This class, which was just a collection of static methods, has been '
+            'deprecated in favor of BinaryMessenger, and it\'s default '
```
`it's` => `its`

hterkelsen replied:
Done.

Hixie commented on packages/flutter/lib/src/services/platform_messages.dart:
```diff
@@ -6,95 +6,50 @@ import 'dart:async';
-typedef _MessageHandler = Future<ByteData> Function(ByteData message);
-
 /// Sends binary messages to and receives binary messages from platform plugins.
 ///
```
probably worth adding something here about how one should use `defaultBinaryMessenger` now instead.

hterkelsen replied:
Done.

Hixie commented:
I like the approach in this patch.

Please follow the breaking change policy (https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes).

Hixie commented:
I'll defer to @amirh, @dnfield, or @goderbauer for final review.

harryterkelsen committed:
Respond to review comments.

- Make the message handler typedef public.
- Fix typos in comments.
- Make clear that defaultBinaryMessenger has replaced BinaryMessages

hterkelsen commented:
I tried to send an email to `flutter-announce@googlegroups.com` but it seems I don't have permission to post there.

I tried to send the following:

As part of a change to generalize Platform Channels so they may be used in Web Plugins, I intend to deprecate BinaryMessages.

Justification

The platform channels use BinaryMessages to send messages from the application to the platform side. The proposed change
will abstract the "binary messenger" in the platform channels such that an implementation which sends messages from the
platform side back to the application can use the same platform channel classes (MethodChannel, OptionalMethodChannel, etc).

Part of this change is to replace the collection of static methods in BinaryMessages with an interface for sending messages,
called BinaryMessenger, and a default instance of BinaryMessenger, called defaultBinaryMessenger.

Links to Issue and PR

Issue: https://github.com/flutter/flutter/issues/29629
PR: https://github.com/flutter/flutter/pull/30406

Migration Steps

defaultBinaryMessenger is a drop-in replacement for BinaryMessages. Simply find and replace every instance of BinaryMessages with defaultBinaryMessages.

Before:

BinaryMessages.setMessageHandler('my-channel', myHandler);

After:

defaultBinaryMessenger.setMessageHandler('my-channel', myHandler);

Conclusion

Please comment on the issue or contact me directly if this change will cause you any issues.

Thanks!

Hixie commented:
cc @kf6gpe for help with flutter-announce.

kf6gpe commented:
@hterkelsen Thanks for the pending contribution! I can give you posting rights to flutter-announce --- reach out to me at the email on my github page and we'll get you set up. Thanks!

dnfield commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -29,23 +29,29 @@ import 'platform_messages.dart';
-  /// Neither [name] nor [codec] may be null.
-  const BasicMessageChannel(this.name, this.codec);
+  /// None of [name], [codec], or [binaryMessenger] may be null.
+  const BasicMessageChannel(this.name, this.codec, [ this.binaryMessenger = defaultBinaryMessenger ])
```
Can we make this a named parameter rather than an optional one?  e.g. `{ this.binaryMessenger = defaultBinaryMessenger }`?  This ends up getting ugly/confusing if we have to add more parameters later.

hterkelsen replied:
Yes. I did it this way because `MethodChannel` has an optional argument for `this.codec` so I could only add `binaryMessenger` as another optional argument.

Hixie commented:
I believe it should be api.flutter.dev.

On Wed, May 22, 2019 at 9:45 AM Dan Field <notifications@github.com> wrote:

> *@dnfield* commented on this pull request.
> ------------------------------
>
> In packages/flutter/lib/src/services/platform_channel.dart
> <https://github.com/flutter/flutter/pull/30406#discussion_r286585302>:
>
> > @@ -287,12 +301,12 @@ class MethodChannel {
>    ///    [StandardMethodCodec].
>    ///  * [JSONMessageCodec] which defines the payload values supported by
>    ///    [JSONMethodCodec].
> -  ///  * <https://docs.flutter.io/javadoc/io/flutter/plugin/common/MethodCall.html>
> +  ///  * <https://docs.flutter.dev/javadoc/io/flutter/plugin/common/MethodCall.html>
>
> Is this ready? Last I heard the docs wasn't ready for .dev links.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/flutter/pull/30406?email_source=notifications&email_token=AAEGSHG3QWETXIVLNUMA4MDPWV2CBA5CNFSM4HDE3AP2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZMZQJA#pullrequestreview-240752676>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEGSHBGO7V5PZP6X7APU7TPWV2CBANCNFSM4HDE3APQ>
> .
>


-- 
Ian Hickson

kf6gpe commented:
We're ready for links to api.flutter.dev at this point, yes. Thanks!



On Wed, May 22, 2019 at 10:22 AM Ian Hickson <notifications@github.com>
wrote:

> I believe it should be api.flutter.dev.
>
> On Wed, May 22, 2019 at 9:45 AM Dan Field <notifications@github.com>
> wrote:
>
> > *@dnfield* commented on this pull request.
> > ------------------------------
> >
> > In packages/flutter/lib/src/services/platform_channel.dart
> > <https://github.com/flutter/flutter/pull/30406#discussion_r286585302>:
> >
> > > @@ -287,12 +301,12 @@ class MethodChannel {
> > /// [StandardMethodCodec].
> > /// * [JSONMessageCodec] which defines the payload values supported by
> > /// [JSONMethodCodec].
> > - /// * <
> https://docs.flutter.io/javadoc/io/flutter/plugin/common/MethodCall.html>
> > + /// * <
> https://docs.flutter.dev/javadoc/io/flutter/plugin/common/MethodCall.html>
> >
> > Is this ready? Last I heard the docs wasn't ready for .dev links.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/flutter/flutter/pull/30406?email_source=notifications&email_token=AAEGSHG3QWETXIVLNUMA4MDPWV2CBA5CNFSM4HDE3AP2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZMZQJA#pullrequestreview-240752676
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAEGSHBGO7V5PZP6X7APU7TPWV2CBANCNFSM4HDE3APQ
> >
> > .
> >
>
>
> --
> Ian Hickson
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/flutter/pull/30406?email_source=notifications&email_token=AAK6QRDYFM4OQTHVGPE42JTPWV6M3A5CNFSM4HDE3AP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV7XI4Q#issuecomment-494892146>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAK6QRAYQFZXSXL3PPDPGKLPWV6M3ANCNFSM4HDE3APQ>
> .
>

hterkelsen commented on packages/flutter/lib/src/services/platform_channel.dart:
```diff
@@ -287,12 +301,12 @@ class MethodChannel {
   ///  * [JSONMessageCodec] which defines the payload values supported by
   ///    [JSONMethodCodec].
-  ///  * <https://docs.flutter.io/javadoc/io/flutter/plugin/common/MethodCall.html>
+  ///  * <https://docs.flutter.dev/javadoc/io/flutter/plugin/common/MethodCall.html>
```
Changed to api.flutter.dev

harryterkelsen committed:
- update url for Flutter javadocs
- make binaryMessenger a named argument in BasicMessageChannel

harryterkelsen committed:
Merge branch 'master' into flutter_channels_2

hterkelsen commented:
Thanks, @dnfield. PTAL

dnfield reviewed (approved):
LGTM

hterkelsen commented:
Thanks! I'll land this tomorrow once the build is fixed and my breaking change announcement has been posted for 24 hours.

harryterkelsen committed:
Merge branch 'master' into flutter_channels_2
